A crash occured after build with dexcount-gradle-plugin

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':teacher:countDebugDexMethods'.
> org.codehaus.groovy.runtime.DefaultGroovyMethods.each(Ljava/util/Collection;Lgroovy/lang/Closure;)Ljava/util/Collection;

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':teacher:countDebugDexMethods'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:42)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.AbstractTask.executeWithoutThrowingTaskFailure(AbstractTask.java:305)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.executeTask(AbstractTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:63)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:51)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:23)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:88)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:29)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:62)
        at org.gradle.execution.DefaultBuildExecuter.access$200(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$2.proceed(DefaultBuildExecuter.java:68)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:62)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:55)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:149)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:106)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:86)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:80)
        at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.java:33)
        at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.java:24)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:36)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
        at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:51)
        at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:171)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:237)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:210)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:35)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:24)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:206)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:169)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
        at org.gradle.launcher.Main.doAction(Main.java:33)
        at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:54)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:35)
        at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
        at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:33)
        at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:130)
        at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:48)
Caused by: java.lang.NoSuchMethodError: org.codehaus.groovy.runtime.DefaultGroovyMethods.each(Ljava/util/Collection;Lgroovy/lang/Closure;)Ljava/util/Collection;
        at com.getkeepsafe.dexcount.PackageTree.printPackageList(PackageTree.groovy:156)
        at com.getkeepsafe.dexcount.PackageTree.print(PackageTree.groovy:125)
        at com.getkeepsafe.dexcount.PackageTree$print$1.call(Unknown Source)
        at com.getkeepsafe.dexcount.DexMethodCountTask.print(DexMethodCountTask.groovy:215)
        at com.getkeepsafe.dexcount.DexMethodCountTask$_printFullTree_closure7.doCall(DexMethodCountTask.groovy:170)
        at com.getkeepsafe.dexcount.DexMethodCountTask$_printToFile_closure11.doCall(DexMethodCountTask.groovy:236)
        at com.getkeepsafe.dexcount.DexMethodCountTask.printToFile(DexMethodCountTask.groovy:234)
        at com.getkeepsafe.dexcount.DexMethodCountTask.printFullTree(DexMethodCountTask.groovy:169)
        at com.getkeepsafe.dexcount.DexMethodCountTask.countMethods(DexMethodCountTask.groovy:72)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:63)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.doExecute(AnnotationProcessingTaskFactory.java:218)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:211)
        at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:200)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:579)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:562)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
        ... 47 more


BUILD FAILED

Total time: 1 mins 2.219 secs
localhost%

I also see crash after updating to Gradle 2.14. Not sure if it is connected to the crash above:
```
Caused by: java.lang.NoClassDefFoundError: org/gradle/logging/ShowStacktrace
        at com.getkeepsafe.dexcount.DexMethodCountPlugin$_applyAndroid_closure1$_closure2.doCall(DexMethodCountPlugin.groovy:62)
        at com.getkeepsafe.dexcount.DexMethodCountPlugin$_applyAndroid_closure1.doCall(DexMethodCountPlugin.groovy:48)
        at org.gradle.api.internal.ClosureBackedAction.execute(ClosureBackedAction.java:67)
        at org.gradle.listener.ActionBroadcast.execute(ActionBroadcast.java:39)
        at org.gradle.api.internal.DefaultDomainObjectCollection.doAdd(DefaultDomainObjectCollection.java:165)
        at org.gradle.api.internal.DefaultDomainObjectCollection.add(DefaultDomainObjectCollection.java:159)
        at com.android.build.gradle.AppExtension.addVariant(AppExtension.java:47)
        at com.android.build.gradle.internal.ApiObjectFactory.create(ApiObjectFactory.java:113)
        at com.android.build.gradle.BasePlugin$12.call(BasePlugin.java:661)
        at com.android.build.gradle.BasePlugin$12.call(BasePlugin.java:654)
        at com.android.builder.profile.ThreadRecorder$1.record(ThreadRecorder.java:55)
        at com.android.builder.profile.ThreadRecorder$1.record(ThreadRecorder.java:47)
        at com.android.build.gradle.BasePlugin.createAndroidTasks(BasePlugin.java:653)
        at com.android.build.gradle.BasePlugin$10$1.call(BasePlugin.java:572)
        at com.android.build.gradle.BasePlugin$10$1.call(BasePlugin.java:569)
        at com.android.builder.profile.ThreadRecorder$1.record(ThreadRecorder.java:55)
        at com.android.builder.profile.ThreadRecorder$1.record(ThreadRecorder.java:47)
        at com.android.build.gradle.BasePlugin$10.execute(BasePlugin.java:568)
        at com.android.build.gradle.BasePlugin$10.execute(BasePlugin.java:565)
        at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:93)
        at org.gradle.internal.event.BroadcastDispatch$ActionInvocationHandler.dispatch(BroadcastDispatch.java:82)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:44)
        ... 70 more
```

--- a/src/main/groovy/com/getkeepsafe/dexcount/PackageTree.groovy
+++ b/src/main/groovy/com/getkeepsafe/dexcount/PackageTree.groovy
@@ -22,6 +22,8 @@ import com.android.dexdeps.MethodRef
 import com.android.dexdeps.Output
 import com.google.gson.stream.JsonWriter
 import groovy.transform.CompileStatic
+import groovy.transform.stc.ClosureParams
+import groovy.transform.stc.FirstParam
 
 import java.nio.CharBuffer
 
@@ -153,7 +155,7 @@ class PackageTree {
             printPackageListHeader(out, opts)
         }
 
-        getChildren(opts).each { it -> it.printPackageListRecursively(out, sb, 0, opts) }
+        forEach(getChildren(opts)) { it -> it.printPackageListRecursively(out, sb, 0, opts) }
     }
 
     private static def printPackageListHeader(Appendable out, PrintOptions opts) {
@@ -192,12 +194,12 @@ class PackageTree {
             out.append('\n')
         }
 
-        getChildren(opts).each { PackageTree it -> it.printPackageListRecursively(out, sb, depth + 1, opts) }
+        forEach(getChildren(opts)) { PackageTree it -> it.printPackageListRecursively(out, sb, depth + 1, opts) }
         sb.setLength(len)
     }
 
     void printTree(Appendable out, PrintOptions opts) {
-        getChildren(opts).each { PackageTree it -> it.printTreeRecursively(out, 0, opts) }
+        forEach(getChildren(opts)) { PackageTree it -> it.printTreeRecursively(out, 0, opts) }
     }
 
     private void printTreeRecursively(Appendable out, int indent, PrintOptions opts) {
@@ -234,7 +236,7 @@ class PackageTree {
 
         out.append("\n")
 
-        getChildren(opts).each { PackageTree it -> it.printTreeRecursively(out, indent + 1, opts) }
+        forEach(getChildren(opts)) { PackageTree it -> it.printTreeRecursively(out, indent + 1, opts) }
     }
 
     void printJson(Appendable out, PrintOptions opts) {
@@ -282,7 +284,7 @@ class PackageTree {
         json.name("children")
         json.beginArray()
 
-        getChildren(opts).each { PackageTree it -> it.printJsonRecursively(json, depth + 1, opts) }
+        forEach(getChildren(opts)) { PackageTree it -> it.printJsonRecursively(json, depth + 1, opts) }
 
         json.endArray()
 
@@ -302,7 +304,7 @@ class PackageTree {
 
         out.append("counts:\n")
 
-        getChildren(opts).each { it.printYamlRecursively(out, 0, opts) }
+        forEach(getChildren(opts)) { it.printYamlRecursively(out, 0, opts) }
     }
 
     private void printYamlRecursively(Appendable out, int depth, PrintOptions opts) {
@@ -337,7 +339,7 @@ class PackageTree {
         } else {
             out.append(indentText)
             out.append("children:\n")
-            children.each { PackageTree child -> child.printYamlRecursively(out, depth + 1, opts) }
+            forEach(children) { PackageTree child -> child.printYamlRecursively(out, depth + 1, opts) }
         }
     }
 
@@ -378,5 +380,28 @@ class PackageTree {
             dot = "<unnamed>." + dot
         }
         return dot
+    }
+
+    /**
+     * Iterates through the elements of a collection, applying the given
+     * closure to each element.
+     *
+     * Workaround for old versions of Gradle that do not have the method
+     * {@link org.codehaus.groovy.runtime.DefaultGroovyMethods#each(Collection, Closure)}.
+     *
+     * Without any workaround, projects building using these versions of Groovy,
+     * we will fail with a MethodMissingException.
+     *
+     * This is observed in projects using Gradle 2.2.1, which bundles Groovy 2.3.7.
+     *
+     * @param collection
+     * @param closure
+     */
+    private static <E> void forEach(
+            Collection<E> collection,
+            @ClosureParams(FirstParam.FirstGenericType.class) Closure closure) {
+        for (E element : collection) {
+            closure.call(element);
+        }
     }
 }

